packages/workbox-broadcast-cache-update/BroadcastCacheUpdate.mjs

```diff
@@ -75,24 +148,68 @@ class BroadcastCacheUpdate {
    *
-   * @param {Response} firstResponse First responses to compare.
-   * @param {Response} secondResponse Second responses to compare.
-   * @param {string} url The URL of the updated request.
-   * @param {string} cacheName Name of the cache the responses belong to.
-   * This is included in the message posted on the broadcast channel.
+   * @param {Event} event The navigation fetch event.
+   * @return {Promise}
+   * @private
    */
-  notifyIfUpdated(firstResponse, secondResponse, url, cacheName) {
-    if (!responsesAreSame(
-      firstResponse, secondResponse, this._headersToCheck)) {
-      broadcastUpdate(this._getChannel(), cacheName, url, this._source);
+  _windowReadyOrTimeout(event) {
+    if (!this._navigationEventsDeferreds.has(event)) {
+      const deferred = new Deferred();
+
+      // Set the deferred on the `_navigationEventsDeferreds` map so it will
+      // be resolved when the next ready message event comes.
+      this._navigationEventsDeferreds.set(event, deferred);
+
+      // But don't wait too long for the message since it may never come.
+      const timeout = setTimeout(() => {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.debug(`Timed out after ${this._deferNoticationTimeout}` +
+              `ms waiting for message from window`);
+        }
+        deferred.resolve();
+      }, this._deferNoticationTimeout);
+
+      // Ensure the timeout is cleared if the deferred promise is resolved.
+      deferred.promise.then(() => clearTimeout(timeout));
     }
+    return this._navigationEventsDeferreds.get(event).promise;
+  }
+
+  /**
+   * Creates a mapping between navigation fetch events and deferreds, and adds
+   * a listener for message events from the window. When message events arrive,
+   * all deferreds in the mapping are resolved.
+   *
+   * Note: it would be easier if we could only resolve the deferred of
+   * navigation fetch event whose client ID matched the source ID of the
+   * message event, but currently client IDs are not exposed on navigation
+   * fetch events: https://www.chromestatus.com/feature/4846038800138240
+   */
+  _initWindowReadyDeferreds() {
+    // A mapping between navigation events and their deferreds.
+    this._navigationEventsDeferreds = new Map();
+
+    // The message listener needs to be added in the initial run of the
+    // service worker, but since we don't actually need to be listening for
+    // messages until the cache updates, we only invoke the callback if set.
+    self.addEventListener('message', (event) => {
+      if (event.data.type === 'WINDOW_READY' &&
+          event.data.meta === 'workbox-window' &&
+          this._navigationEventsDeferreds.size > 0) {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.debug(`Received WINDOW_READY event: `, event);
+        }
+        // Resolve any pending deferreds.
+        for (const [, deferred] of this._navigationEventsDeferreds.entries()) {
```

Comment:
Can you switch this to

```
for (const deferred of this._navigationEventsDeferreds.values())
```

Using `entries()` and then ignoring the first item looks weird.

Reply:
Ahhh, why didn't I think of that! :)